### DISCUSSION TITLE
Add a way to change bootstrap install link

### DISCUSSION THREAD

**Discussion by Maxython** [termux/termux-app#2905] [Ideas]
In the termux application code, a link for installing bootstrap is immediately installed, which cannot be changed in any way in the termux application itself. For termux users, there are two options for installing non-original bootstrap. 
 - The first one is a self-installation of bootsrap, which is one-time (if the user reinstalls bootstrap, then what he installed will be removed and bootstrap will be installed from the link in the code) and has its own nuances. 
 - The second is to compile the termux application with a modified link to install bootsrap (this option for users I think will be very difficult). 

So why not add a way to change the bootstrap install link?  
**Implementation example:**  
There will be a file in the application, let's say it's called `bootstrapLink`, which is not in `PREFIX`, but in the termux `files` directory. The `bootstrapLink` file contains a link to the bootsrap installation. The termux application reads the `bootstrapLink` file and installs bootsrap from the file link. If the `bootstrapLink` file has an invalid reference, or the file itself does not exist, then the termux application recreates the `bootstrapLink` file with the reference specified in the termux application code.

**Comment by agnostic-apollo** [comment]
`termux-app` runs code depending on the bootstrap that comes in the apk. If user changes the package manager, app will not run correctly. However, the bootstraps built could include the package variant as well in a config file, so if custom one downloaded or loaded from local filesystem, it can be compared against the one in the apk.

>let's say it's called bootstrapLink, which is not in PREFIX, but in the termux files directory

The file could not exist by default and apk bootstrap is used. If it does, it can be checked if it contains absolute file path or https url to bootstrap, if failed to download/load, could just fallback to apk bootstrap.

Although, I plan on adding a prompt before installing bootstrap instead of automatically installing, it will have advance options, so could just add a field for custom bootstrap link. This way user would have to manually enter in a UI instead of background and should be safer from malicious scripts creating file and deleting prefix to have persistent malicious bootstrap unless app is uninstalled.

https://github.com/termux/termux-app/blob/211340781bcb0576ffe100fda2e0ed87d7333d53/app/build.gradle#L6-L12

**Reply by Maxython** [reply]
Excuse my stupidity, but what do you mean by `apk`? Is this what android system or file format?

**Reply by agnostic-apollo** [reply]
Bootstraps are not downloaded from the internet at runtime with a link, they are included in each apk in `lib/<arch>/libtermux-bootstrap.so`.

https://github.com/termux/termux-app/commit/c76cec1

**Reply by Maxython** [reply]
You just broke my idea of ​​how bootstrap is installed (well, or I misunderstood something). Can you describe the bootstrap installation process when termux is started?  

PS: if bootstrap is already in the apk file, then where is it stored (directory coordinates)?

**Reply by agnostic-apollo** [reply]
Check https://github.com/termux/termux-packages/wiki/For-maintainers#bootstraps

**Reply by Maxython** [reply]
Thank

**Reply by Maxython** [reply]
one question: why is the termux application compiled with the bootstrap archive? Wouldn't it be easier to download it from the Internet? If this is somehow related to security, then we can create an archive (service, something on the other side with packages but with bootstraps) that are officially verified by the termux developers.  

(apparently I went too deep with bootstrap ideas ._.)

**Reply by ghost** [reply]
It was previously using network-based bootstrap installation in the past versions of Termux, but there are issues regarding that such as "Network connectivity issues" 

https://github.com/termux/termux-packages/issues/1761

**Reply by Maxython** [reply]
Ok, i see

**Comment by Maxython** [comment]
Well, change of idea. Can you add pacman-based bootsrap to termux app? Yes, the original idea came from that. There is simply no "convenient" way to change the package manager to termux. Yes, can reinstall all packages with pacman, but there are pitfalls. Yes, can install this bootstrap on termux yourself, but as I mentioned this is a one-time option. (sorry if this sounded rude)

**Reply by agnostic-apollo** [reply]
>Can you add pacman-based bootsrap to termux app? 

I plan on adding it. Have to fix bootstrap stuff in `termux-packages` as well and merge related pull requests.

Package manager  should usually not be mixed on distros either due to incompatibilities and the one that comes by default should be used, so having same restriction makes sense app wise too. If users want to use `pacman` manager and `pacman` packages from non-official 3rd party repos, then they should install the app from said repos too. You can fork `termux-app` and releases can be published for `pacman` manager and users can install from there, doesn't sound too complex.

But I could add support to run app based on bootstrap installed, basically use `build.gradle` value if apk bootstrap is used, and if custom one is provided, then read manager from inside bootstrap zip, then you won't have to publish separate app, only provide users with url and checksums for bootstraps.

>sorry if this sounded rude

Naa...

**Reply by Maxython** [reply]
Note: pacman-based bootstrap is built and found in the releases of the [termux-pacman/termux-packages](https://github.com/termux-pacman/termux-packages/releases) repository.

**Reply by agnostic-apollo** [reply]
Okay, cool.
